From a625449792c05f79729374fbb0fc110600085d41 Mon Sep 17 00:00:00 2001
From: Etienne Carriere <etienne.carriere@foss.st.com>
Date: Mon, 15 Jul 2024 16:37:37 +0200
Subject: [PATCH 1165/1695] scpfw: module: scmi_clock: specifc OSTLv4/v5
 duty-cycle/round_rate messages

Support for STM32MP OSTL legacy SCMI clock protocol messages
when CFG_STM32_OTSL_SCMI_CLOCK_SUPPORT is enabled.

SCMI clock protocol message IDs MOD_SCMI_CLOCK_DUTY_CYCLE_GET and
MOD_SCMI_CLOCK_ROUND_RATE_GET are specific to OSTLv4 and OSTLv5 and
supported only when the SCMI agent uses SMCI Clock protocol v1.0 or v2.0.
This is because they conflict with SCMI clock protocol v3.0 message IDs
MOD_SCMI_CLOCK_CONFIG_GET and MOD_SCMI_CLOCK_CONFIG_GET.

This change implements SCMI clock protocol v3.0 message IDs
MOD_SCMI_CLOCK_CONFIG_SET (new message payload ABI) and
MOD_SCMI_CLOCK_CONFIG_GET (to get the duty cycle of a clock). When
using SCMI clock protocol v3.0, the clock round rate request goes
through MOD_SCMI_CLOCK_DESCRIBE_RATES message ID.

This change makes the SCMI server to treat MOD_SCMI_CLOCK_CONFIG_SET
over SCMI clock protocol v2.0 and OSTL legacy message IDs
MOD_SCMI_CLOCK_DUTY_CYCLE_GET and MOD_SCMI_CLOCK_ROUND_RATE_GET, using
the payload size to distinguish between these OSTL specific messages
and the SCMI clock protocol v3.0 message IDs MOD_SCMI_CLOCK_ROUND_RATE_GET
and MOD_SCMI_CLOCK_POSSIBLE_PARENT_GET.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 80e1b430f9c9eea3474097a599e75a59338f203e scpfw: module: scmi_clock: specifc OSTLv4/v5 duty-cycle/round_rate messages]
Change-Id: Ifcc197d0b9c710dcd070f377a4e7c991021714a3
Signed-off-by: Etienne Carriere <etienne.carriere@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/393550
Tested-by: Thomas BOURGOIN <thomas.bourgoin@foss.st.com>
Domain-Review: Thomas BOURGOIN <thomas.bourgoin@foss.st.com>
Reviewed-by: Thomas BOURGOIN <thomas.bourgoin@foss.st.com>
---
 core/arch/arm/plat-stm32mp1/conf.mk           |   3 +
 core/arch/arm/plat-stm32mp2/conf.mk           |   3 +
 .../module/scmi/include/mod_scmi_std.h        |  16 +-
 .../scmi_clock/include/internal/scmi_clock.h  |  69 +++++-
 .../module/scmi_clock/src/mod_scmi_clock.c    | 213 ++++++++++++++++--
 5 files changed, 276 insertions(+), 28 deletions(-)

diff --git a/core/arch/arm/plat-stm32mp1/conf.mk b/core/arch/arm/plat-stm32mp1/conf.mk
index 69deec779..cb2911a59 100644
--- a/core/arch/arm/plat-stm32mp1/conf.mk
+++ b/core/arch/arm/plat-stm32mp1/conf.mk
@@ -450,6 +450,9 @@ $(call force,CFG_DRIVERS_RTC,y)
 $(call force,CFG_RTC_PTA,y)
 endif
 
+# Support for STM32MP OSTL legacy SCMI messages
+CFG_STM32_OTSL_SCMI_CLOCK_SUPPORT ?= n
+
 # Enable Early TA NVMEM for provisioning management
 CFG_TA_STM32MP_NVMEM ?= y
 ifeq ($(CFG_TA_STM32MP_NVMEM),y)
diff --git a/core/arch/arm/plat-stm32mp2/conf.mk b/core/arch/arm/plat-stm32mp2/conf.mk
index 7d1891621..c60eceee6 100644
--- a/core/arch/arm/plat-stm32mp2/conf.mk
+++ b/core/arch/arm/plat-stm32mp2/conf.mk
@@ -332,3 +332,6 @@ $(call force,CFG_STM32_LTDC,y)
 CFG_TUI_FRAME_BUFFER_SIZE_MAX ?= 0x01000000
 CFG_RESERVED_VASPACE_SIZE ?= (10 * 1024 * 1024 + $(CFG_TUI_FRAME_BUFFER_SIZE_MAX))
 endif
+
+# Support for STM32MP OSTL legacy SCMI messages
+CFG_STM32_OTSL_SCMI_CLOCK_SUPPORT ?= y
diff --git a/core/lib/scmi-server/SCP-firmware/module/scmi/include/mod_scmi_std.h b/core/lib/scmi-server/SCP-firmware/module/scmi/include/mod_scmi_std.h
index 1724406bf..454b608ee 100644
--- a/core/lib/scmi-server/SCP-firmware/module/scmi/include/mod_scmi_std.h
+++ b/core/lib/scmi-server/SCP-firmware/module/scmi/include/mod_scmi_std.h
@@ -156,19 +156,31 @@ enum scmi_perf_command_id {
  * \brief SCMI Clock Protocol Message IDs
  */
 enum scmi_clock_command_id {
+    /* Commands introduced in SCMI clock protocol v1.0 */
     MOD_SCMI_CLOCK_ATTRIBUTES = 0x003,
     MOD_SCMI_CLOCK_DESCRIBE_RATES = 0x004,
     MOD_SCMI_CLOCK_RATE_SET = 0x005,
     MOD_SCMI_CLOCK_RATE_GET = 0x006,
     MOD_SCMI_CLOCK_CONFIG_SET = 0x007,
+    /* Commands introduced in SCMI clock protocol v2.0 */
     MOD_SCMI_CLOCK_NAME_GET = 0x008,
     MOD_SCMI_CLOCK_RATE_NOTIFY = 0x009,
     MOD_SCMI_CLOCK_RATE_CHANGE_REQUESTED_NOTIFY = 0x00A,
-    MOD_SCMI_CLOCK_DUTY_CYCLE_GET = 0x00B,
-    MOD_SCMI_CLOCK_ROUND_RATE_GET = 0x00C,
+    /* Commands introduced in SCMI clock protocol v3.0 */
+    MOD_SCMI_CLOCK_CONFIG_GET = 0x00B,
+    MOD_SCMI_CLOCK_POSSIBLE_PARENT_GET = 0x00c,
+    MOD_SCMI_CLOCK_PARENT_SET = 0x00d,
+    MOD_SCMI_CLOCK_PARENT_GET = 0x00e,
+    MOD_SCMI_CLOCK_GET_PERMISSIONS = 0x00f,
     MOD_SCMI_CLOCK_COMMAND_COUNT
 };
 
+#ifdef CFG_STM32_OTSL_SCMI_CLOCK_SUPPORT
+/* OSTLv4/v5 specific commands (deprecated since OSTLv6) */
+#define MOD_SCMI_CLOCK_DUTY_CYCLE_GET	0x00B
+#define MOD_SCMI_CLOCK_ROUND_RATE_GET	0x00C
+#endif
+
 /*!
  * \brief SCMI Sensor Protocol
  */
diff --git a/core/lib/scmi-server/SCP-firmware/module/scmi_clock/include/internal/scmi_clock.h b/core/lib/scmi-server/SCP-firmware/module/scmi_clock/include/internal/scmi_clock.h
index 7671e0e8f..e52eccfca 100644
--- a/core/lib/scmi-server/SCP-firmware/module/scmi_clock/include/internal/scmi_clock.h
+++ b/core/lib/scmi-server/SCP-firmware/module/scmi_clock/include/internal/scmi_clock.h
@@ -13,7 +13,7 @@
 
 #include <mod_clock.h>
 
-#define SCMI_PROTOCOL_VERSION_CLOCK UINT32_C(0x10000)
+#define SCMI_PROTOCOL_VERSION_CLOCK UINT32_C(0x30000)
 
 /*
  * Identifiers for the type of request being processed
@@ -182,11 +182,36 @@ struct scmi_clock_rate_set_p2a {
 #define SCMI_CLOCK_CONFIG_SET_ENABLE_MASK \
     (UINT32_C(0x1) << SCMI_CLOCK_CONFIG_SET_ENABLE_POS)
 
+#define SCMI_CLOCK_CONFIG_SET_EXT_TYPE_MASK		0x00FF0000
+#define SCMI_CLOCK_CONFIG_SET_EXT_TYPE_POS		16
+
+#define SCMI_CLOCK_EXTENDED_CONFIG_TYPE_UNUSED		0
+#define SCMI_CLOCK_EXTENDED_CONFIG_TYPE_DUTY_CYCLE	1
+#define SCMI_CLOCK_EXTENDED_CONFIG_TYPE_PHASE		2
+
+#define SCMI_CLOCK_CONFIG_SET_EXT_TYPE_UNUSED \
+    (SCMI_CLOCK_EXTENDED_CONFIG_TYPE_UNUSED << \
+     SCMI_CLOCK_CONFIG_SET_EXT_TYPE_POS)
+
+#define SCMI_CLOCK_CONFIG_SET_EXT_TYPE_DUTY_CYCLE \
+    (SCMI_CLOCK_EXTENDED_CONFIG_TYPE_DUTY_CYCLE << \
+     SCMI_CLOCK_CONFIG_SET_EXT_TYPE_POS)
+
+#define SCMI_CLOCK_CONFIG_SET_EXT_TYPE_PHASE \
+    (SCMI_CLOCK_EXTENDED_CONFIG_TYPE_PHASE << \
+     SCMI_CLOCK_CONFIG_SET_EXT_TYPE_POS)
+
 struct scmi_clock_config_set_a2p {
     uint32_t clock_id;
     uint32_t attributes;
 };
 
+struct scmi_clock_config_set_v2_a2p {
+    uint32_t clock_id;
+    uint32_t attributes;
+    uint32_t extended_config_val;
+};
+
 struct scmi_clock_config_set_p2a {
     int32_t status;
 };
@@ -287,32 +312,62 @@ struct scmi_clock_rate_change_request_notify_p2a {
 };
 
 /*
- * Clock get duty cycle
+ * Clock get config
+ */
+#define SCMI_CLOCK_CONFIG_GET_EXT_TYPE_MASK		GENMASK_32(7, 0)
+
+struct scmi_clock_config_get_a2p {
+    uint32_t clock_id;
+    uint32_t flags;
+};
+
+struct scmi_clock_config_get_p2a {
+    int32_t status;
+    uint32_t attributes;
+    uint32_t config;
+    uint32_t extended_config_val;
+};
+
+#ifdef CFG_STM32_OTSL_SCMI_CLOCK_SUPPORT
+/*
+ * OSTLv4/v5 Clock get duty cycle message (deprecated)
  */
 
-struct scmi_clock_duty_cycle_get_a2p {
+struct scmi_clock_ostl_duty_cycle_get_a2p {
     uint32_t clock_id;
 };
 
-struct scmi_clock_duty_cycle_get_p2a {
+struct scmi_clock_ostl_duty_cycle_get_p2a {
     int32_t status;
     uint32_t numerator;
     uint32_t denominator;
 };
 
 /*
- * Clock get round rate
+ * OSTLv4/v5 Clock get round rate messsage (deprecated)
  */
 
-struct scmi_clock_round_rate_get_a2p {
+struct scmi_clock_ostl_round_rate_get_a2p {
     uint32_t flags;
     uint32_t clock_id;
     uint32_t rate[2];
 };
 
-struct scmi_clock_round_rate_get_p2a {
+struct scmi_clock_ostl_round_rate_get_p2a {
     int32_t status;
     uint32_t rate[2];
 };
 
+/*
+ * Clock possible parent
+ * Currently not supported but input message structure is defined
+ * to address OSTLv4/v5 SCMI_CLOCK_ROUND_RATE_GET command issue.
+ */
+
+struct scmi_clock_possible_parent_a2p {
+    uint32_t clock_id;
+    uint32_t skip_parents;
+};
+
+#endif
 #endif /* INTERNAL_SCMI_CLOCK_H */
diff --git a/core/lib/scmi-server/SCP-firmware/module/scmi_clock/src/mod_scmi_clock.c b/core/lib/scmi-server/SCP-firmware/module/scmi_clock/src/mod_scmi_clock.c
index 5ee7a38f4..01f2627cc 100644
--- a/core/lib/scmi-server/SCP-firmware/module/scmi_clock/src/mod_scmi_clock.c
+++ b/core/lib/scmi-server/SCP-firmware/module/scmi_clock/src/mod_scmi_clock.c
@@ -126,6 +126,8 @@ static int scmi_clock_rate_set_handler(fwk_id_t service_id,
     const uint32_t *payload);
 static int scmi_clock_config_set_handler(fwk_id_t service_id,
     const uint32_t *payload);
+static int scmi_clock_config_set_v2_handler(fwk_id_t service_id,
+    const uint32_t *payload);
 static int scmi_clock_describe_rates_handler(fwk_id_t service_id,
     const uint32_t *payload);
 static int scmi_clock_name_get_handler(fwk_id_t service_id,
@@ -134,10 +136,16 @@ static int scmi_clock_rate_notify_handler(fwk_id_t service_id,
     const uint32_t *payload);
 static int scmi_clock_rate_change_request_notify_handler(fwk_id_t service_id,
     const uint32_t *payload);
-static int scmi_clock_duty_cycle_get_handler(fwk_id_t service_id,
+static int scmi_clock_config_get_handler(fwk_id_t service_id,
+    const uint32_t *payload);
+
+#ifdef CFG_STM32_OTSL_SCMI_CLOCK_SUPPORT
+/* Handler for OSTLv4/v5 specific message IDs */
+static int scmi_clock_ostl_duty_cycle_get_handler(fwk_id_t service_id,
     const uint32_t *payload);
-static int scmi_clock_round_rate_get_handler(fwk_id_t service_id,
+static int scmi_clock_ostl_round_rate_get_handler(fwk_id_t service_id,
     const uint32_t *payload);
+#endif
 
 /*
  * Internal variables.
@@ -154,13 +162,12 @@ static int (*const handler_table[MOD_SCMI_CLOCK_COMMAND_COUNT])(
     [MOD_SCMI_CLOCK_ATTRIBUTES] = scmi_clock_attributes_handler,
     [MOD_SCMI_CLOCK_RATE_GET] = scmi_clock_rate_get_handler,
     [MOD_SCMI_CLOCK_RATE_SET] = scmi_clock_rate_set_handler,
-    [MOD_SCMI_CLOCK_CONFIG_SET] = scmi_clock_config_set_handler,
+    [MOD_SCMI_CLOCK_CONFIG_SET] = scmi_clock_config_set_v2_handler,
     [MOD_SCMI_CLOCK_DESCRIBE_RATES] = scmi_clock_describe_rates_handler,
     [MOD_SCMI_CLOCK_NAME_GET] = scmi_clock_name_get_handler,
     [MOD_SCMI_CLOCK_RATE_NOTIFY] = scmi_clock_rate_notify_handler,
     [MOD_SCMI_CLOCK_RATE_CHANGE_REQUESTED_NOTIFY] = scmi_clock_rate_change_request_notify_handler,
-    [MOD_SCMI_CLOCK_DUTY_CYCLE_GET] = scmi_clock_duty_cycle_get_handler,
-    [MOD_SCMI_CLOCK_ROUND_RATE_GET] = scmi_clock_round_rate_get_handler,
+    [MOD_SCMI_CLOCK_CONFIG_GET] = scmi_clock_config_get_handler,
 };
 
 static const unsigned int payload_size_table[MOD_SCMI_CLOCK_COMMAND_COUNT] = {
@@ -175,7 +182,7 @@ static const unsigned int payload_size_table[MOD_SCMI_CLOCK_COMMAND_COUNT] = {
     [MOD_SCMI_CLOCK_RATE_SET] =
         (unsigned int)sizeof(struct scmi_clock_rate_set_a2p),
     [MOD_SCMI_CLOCK_CONFIG_SET] =
-        (unsigned int)sizeof(struct scmi_clock_config_set_a2p),
+        (unsigned int)sizeof(struct scmi_clock_config_set_v2_a2p),
     [MOD_SCMI_CLOCK_DESCRIBE_RATES] =
         (unsigned int)sizeof(struct scmi_clock_describe_rates_a2p),
     [MOD_SCMI_CLOCK_NAME_GET] =
@@ -184,10 +191,8 @@ static const unsigned int payload_size_table[MOD_SCMI_CLOCK_COMMAND_COUNT] = {
         (unsigned int)sizeof(struct scmi_clock_rate_notify_a2p),
     [MOD_SCMI_CLOCK_RATE_CHANGE_REQUESTED_NOTIFY] =
         (unsigned int)sizeof(struct scmi_clock_rate_change_request_notify_a2p),
-    [MOD_SCMI_CLOCK_DUTY_CYCLE_GET] =
-        (unsigned int)sizeof(struct scmi_clock_duty_cycle_get_a2p),
-    [MOD_SCMI_CLOCK_ROUND_RATE_GET] =
-        (unsigned int)sizeof(struct scmi_clock_round_rate_get_a2p),
+    [MOD_SCMI_CLOCK_CONFIG_GET] =
+        (unsigned int)sizeof(struct scmi_clock_config_get_a2p),
 };
 
 /*
@@ -1130,6 +1135,8 @@ exit:
 /*
  * Clock Config Set
  */
+
+/* Handler for CONFIG_SET message for SCMI clock protocol v1.0 and v2.0 */
 static int scmi_clock_config_set_handler(fwk_id_t service_id,
     const uint32_t *payload)
 {
@@ -1139,7 +1146,7 @@ static int scmi_clock_config_set_handler(fwk_id_t service_id,
     const struct scmi_clock_config_set_a2p *parameters;
     const struct mod_scmi_clock_device *clock_device;
     unsigned int agent_id;
-    struct scmi_clock_rate_set_p2a return_values = {
+    struct scmi_clock_config_set_p2a return_values = {
         .status = (int32_t)SCMI_GENERIC_ERROR
     };
     enum mod_scmi_clock_policy_status policy_status;
@@ -1222,6 +1229,123 @@ exit:
         service_id, &return_values, response_size);
 }
 
+/*
+ * Handler for CONFIG_SET message for SCMI clock protocol v3.0 and later.
+ * Only support clock enable/disable over CONFIG_SET message.
+ */
+static int scmi_clock_config_set_v2_handler(fwk_id_t service_id,
+    const uint32_t *payload)
+{
+    const struct scmi_clock_config_set_v2_a2p *parameters;
+    struct scmi_clock_config_set_p2a return_values = {
+        .status = (int32_t)SCMI_GENERIC_ERROR
+    };
+
+    parameters = (const struct scmi_clock_config_set_v2_a2p*)payload;
+
+    switch (parameters->attributes & SCMI_CLOCK_CONFIG_SET_EXT_TYPE_MASK) {
+    case SCMI_CLOCK_CONFIG_SET_EXT_TYPE_UNUSED:
+        return scmi_clock_config_set_handler(service_id, payload);
+
+    default:
+        FWK_LOG_DEBUG("[SCMI-CLK] Extended type %d is not supported",
+            (parameters->attributes & SCMI_CLOCK_CONFIG_SET_EXT_TYPE_MASK) >>
+            SCMI_CLOCK_CONFIG_SET_EXT_TYPE_POS);
+
+        return_values.status = (int32_t)SCMI_NOT_SUPPORTED;
+        return scmi_clock_ctx.scmi_api->respond(
+            service_id,
+            &return_values,
+            sizeof(return_values.status));
+    }
+}
+
+/* Handler for CONFIG_GET message */
+static int scmi_clock_config_get_handler(fwk_id_t service_id,
+    const uint32_t *payload)
+{
+    int status;
+    size_t response_size;
+    unsigned int agent_id;
+    unsigned int clock_dev_idx;
+    uint32_t duty_cycle_numerator;
+    uint32_t duty_cycle_denominator;
+    enum mod_clock_state clock_state;
+    const struct mod_scmi_clock_device *clock_device;
+    const struct scmi_clock_config_get_a2p *parameters;
+    struct scmi_clock_config_get_p2a return_values = {
+        .status = (int32_t)SCMI_GENERIC_ERROR
+    };
+
+    parameters = (const struct scmi_clock_config_get_a2p*)payload;
+
+    status = scmi_clock_get_clock_device_entry(
+        service_id, parameters->clock_id, &clock_device);
+    if (status != FWK_SUCCESS) {
+        return_values.status = (int32_t)SCMI_NOT_FOUND;
+        goto exit;
+    }
+
+#ifdef BUILD_HAS_MOD_RESOURCE_PERMS
+    /* We don't support yet resource permission filtering on config get */
+#endif
+
+    status = scmi_clock_ctx.scmi_api->get_agent_id(service_id, &agent_id);
+    if (status != FWK_SUCCESS) {
+        goto exit;
+    }
+
+    switch (parameters->flags & SCMI_CLOCK_CONFIG_SET_EXT_TYPE_MASK) {
+    case SCMI_CLOCK_CONFIG_SET_EXT_TYPE_UNUSED:
+        /*
+         * Shortcut use of internal messages to get the clock state.
+         */
+        clock_dev_idx = fwk_id_get_element_idx(clock_device->element_id);
+        clock_state = scmi_clock_get_agent_clock_state(
+            agent_id,
+            scmi_clock_ctx.clock_ops[clock_dev_idx].scmi_clock_idx);
+
+        return_values = (struct scmi_clock_config_get_p2a){
+            .status = SCMI_SUCCESS,
+            .attributes = clock_state,
+        };
+        break;
+
+    case SCMI_CLOCK_CONFIG_SET_EXT_TYPE_DUTY_CYCLE:
+        status = scmi_clock_ctx.clock_api->get_duty_cycle(
+            clock_device->element_id,
+            &duty_cycle_numerator,
+            &duty_cycle_denominator);
+        if (status == FWK_E_SUPPORT) {
+            return_values.status = (int32_t)SCMI_NOT_SUPPORTED;
+        } else if (status != FWK_SUCCESS) {
+            return_values.status = (int32_t)SCMI_GENERIC_ERROR;
+        } else {
+            return_values = (struct scmi_clock_config_get_p2a){
+                .status = SCMI_SUCCESS,
+                .extended_config_val =
+                    ((unsigned long long)duty_cycle_numerator * 100) /
+                    duty_cycle_denominator,
+            };
+        }
+        break;
+
+    default:
+        FWK_LOG_DEBUG("[SCMI-CLK] Extended type %d is not supported",
+            (parameters->flags & SCMI_CLOCK_CONFIG_SET_EXT_TYPE_MASK) >>
+            SCMI_CLOCK_CONFIG_SET_EXT_TYPE_POS);
+
+        return_values.status = (int32_t)SCMI_NOT_SUPPORTED;
+        break;
+    }
+
+exit:
+    response_size = (return_values.status == SCMI_SUCCESS) ?
+        sizeof(return_values) : sizeof(return_values.status);
+    return scmi_clock_ctx.scmi_api->respond(
+        service_id, &return_values, response_size);
+}
+
 /*
  * Clock Describe Rates
  */
@@ -1484,12 +1608,16 @@ static int scmi_clock_rate_change_request_notify_handler(fwk_id_t service_id,
     return FWK_SUCCESS;
 }
 
-static int scmi_clock_duty_cycle_get_handler(fwk_id_t service_id,
+#ifdef CFG_STM32_OTSL_SCMI_CLOCK_SUPPORT
+/*
+ * Handler for OSTLv4/v5 specific CLOCK_DUTY_CYCLE_GET message ID
+ */
+static int scmi_clock_ostl_duty_cycle_get_handler(fwk_id_t service_id,
     const uint32_t *payload)
 
 {
-    const struct scmi_clock_duty_cycle_get_a2p *parameters;
-    struct scmi_clock_duty_cycle_get_p2a return_values = {
+    const struct scmi_clock_ostl_duty_cycle_get_a2p *parameters;
+    struct scmi_clock_ostl_duty_cycle_get_p2a return_values = {
         .status = (int32_t)SCMI_SUCCESS
     };
     const struct mod_scmi_clock_device *clock_device;
@@ -1497,7 +1625,7 @@ static int scmi_clock_duty_cycle_get_handler(fwk_id_t service_id,
     size_t response_size;
     int status;
 
-    parameters = (const struct scmi_clock_duty_cycle_get_a2p*)payload;
+    parameters = (const struct scmi_clock_ostl_duty_cycle_get_a2p*)payload;
 
     status = scmi_clock_get_clock_device_entry(
         service_id, parameters->clock_id, &clock_device);
@@ -1541,14 +1669,15 @@ exit:
 }
 
 /*
+ * Handler for OSTLv4/v5 specific CLOCK_ROUND_RATE_GET message ID.
  * Use clock set rate with round nearest flag support to
  * get rounded value.
  */
-static int scmi_clock_round_rate_get_handler(fwk_id_t service_id,
+static int scmi_clock_ostl_round_rate_get_handler(fwk_id_t service_id,
     const uint32_t *payload)
 {
-    const struct scmi_clock_round_rate_get_a2p *parameters;
-    struct scmi_clock_round_rate_get_p2a return_values = {
+    const struct scmi_clock_ostl_round_rate_get_a2p *parameters;
+    struct scmi_clock_ostl_round_rate_get_p2a return_values = {
         .status = (int32_t)SCMI_SUCCESS,
     };
     const struct mod_scmi_clock_device *clock_device;
@@ -1557,7 +1686,7 @@ static int scmi_clock_round_rate_get_handler(fwk_id_t service_id,
     size_t response_size;
     int status;
 
-    parameters = (const struct scmi_clock_round_rate_get_a2p*)payload;
+    parameters = (const struct scmi_clock_ostl_round_rate_get_a2p*)payload;
 
     status = scmi_clock_get_clock_device_entry(
         service_id,
@@ -1612,6 +1741,7 @@ exit:
     return scmi_clock_ctx.scmi_api->respond(
         service_id, &return_values, response_size);
 }
+#endif
 
 /*
  * SCMI module -> SCMI clock module interface
@@ -1639,6 +1769,51 @@ static int scmi_clock_message_handler(fwk_id_t protocol_id, fwk_id_t service_id,
         goto error;
     }
 
+    switch (message_id) {
+    case MOD_SCMI_CLOCK_CONFIG_SET:
+        /*
+         * Special case to support CONFIG_SET message ID for SCMI Clock
+         * protocol before v3.0 since SCMI clock protocol v3.0 introduced
+         * an extra input paremters on the message payload.
+         */
+        if (payload_size == sizeof(struct scmi_clock_config_set_a2p))
+            return scmi_clock_config_set_handler(service_id, payload);
+        break;
+
+#ifdef CFG_STM32_OTSL_SCMI_CLOCK_SUPPORT
+    case MOD_SCMI_CLOCK_DUTY_CYCLE_GET:
+        /*
+         * Special case to support OSTLv4/v5 duty-cycle requests:
+         * based on input message size, distinguish standard
+         * SCMI clock protocol v3.0 SCMI_CLOCK_CONFIG_GET
+         * from OSTLv4/v5 SCMI_CLOCK_DUTY_CYCLE_GET.
+         */
+        static_assert(sizeof(struct scmi_clock_config_get_a2p) !=
+                      sizeof(struct scmi_clock_ostl_duty_cycle_get_a2p));
+
+        if (payload_size == sizeof(struct scmi_clock_ostl_duty_cycle_get_a2p))
+            return scmi_clock_ostl_duty_cycle_get_handler(service_id, payload);
+        break;
+
+    case MOD_SCMI_CLOCK_ROUND_RATE_GET:
+        /*
+         * Special case to support OSTLv4/v5 duty-cycle requests:
+         * based on input message size, distinguish standard
+         * SCMI clock protocol v3.0 SCMI_CLOCK_POSSIBLE_PARENT_GET
+         * from OSTLv4/v5 SCMI_CLOCK_ROUND_RATE_GET.
+         */
+        static_assert(sizeof(struct scmi_clock_possible_parent_a2p) !=
+                       sizeof(struct scmi_clock_ostl_round_rate_get_a2p));
+
+        if (payload_size == sizeof(struct scmi_clock_ostl_round_rate_get_a2p))
+            return scmi_clock_ostl_round_rate_get_handler(service_id, payload);
+        break;
+#endif
+
+    default:
+        break;
+    }
+
     if (payload_size != payload_size_table[message_id]) {
         return_value = (int32_t)SCMI_PROTOCOL_ERROR;
         goto error;
-- 
2.47.2

